content/blog/2024/12/how-to-migrate-from-idf-bootloader-to-mcuboot-using-esp-self-reflasher/index.md

```diff
@@ -0,0 +1,271 @@
+
+On the firmware area, there are a variety of reasons that may motivate choosing a flexible platform for a project or actually migrating from a production running RTOS (Real Time Operating System) to another one, they may range from technical to strategic purposes.
+
+Relatively, changing an RTOS may not be as straight forward as a firmware update. One of the concerns that needs to be addressed beforehand is the bootloader compatibility.
+
+Besides initializing the system, the bootloader may be responsible for handling safe and secure firmware updates. However, it typically cannot update itself, either by design or due to device restrictions, such as flash bootloader region protection. Also in most of the scenarios, the device is not physically accessible, thus a direct serial flashing is not practical.
+
+This tutorial demonstrates how ESP Self-Reflasher component can be used to migrate an IDF-bootloader-based application (OTA capable) to a MCUboot-based-bootloader application. The ESP Self-Reflasher component is used on an middle step application for enabling the device to "reflash" its own firmware.
+
+---
+**Note**
+The MCUboot compatible application used in this guide is a NuttX RTOS application, but could be other like a Zephyr RTOS application, for instance.
+
+---
+
+The overall process described in this guide can be illustrated as:
+
+```mermaid
+flowchart LR
+    A[IDF bootloader + IDF application with OTA capabilities] -->|OTA process| B[Reflashing application with ESP Self-Reflasher component]
+    B -->|Reflashing process| C(MCUboot bootloader + NuttX application)
+```
+
+## ESP Self Reflasher component
+
+The [ESP Self-Reflasher](https://components.espressif.com/components/espressif/esp-self-reflasher/versions/1.0.0) is an ESP-IDF component that allows the device to be "reflashed" with new firmware. This is achieved through one of the following ways of operation: downloading the reflashing image or embedding the reflashing image. The component has examples for each that will be used on this guide.
+
+## Environment pre requisites
+
+Besides ESP-IDF ([Get started guide here](https://docs.espressif.com/projects/esp-idf/en/stable/esp32/get-started/index.html)), the following are required:
+
+- ESP Self-Reflasher: clone it to `<IDF_DIR>/components` with Git:
+
+```bash
+cd <IDF_DIR>/components
+git clone https://github.com/espressif/esp-self-reflasher.git
+```
+
+- NuttX: in order to build the application that will be migrated to, set the NuttX workspace. [Get started guide here](https://developer.espressif.com/blog/nuttx-getting-started/).
+
+- MCUboot: in the case of building the bootloader standalone, follow the instructions [here](https://docs.mcuboot.com/readme-espressif.html).
+
+## Step-by-step Guide
+
+### Building the reflashing images (MCUboot and NuttX application)
```

Comment:
The procedures are commonly titled with verbs in their infinitive form to make the title focus on action almost like a command:

- Build the reflash images (MCUboot and NuttX application)
- Build the reflashing application
- Build and flash the IDF application